sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,921 @@
+
+* The function must be declared at namespace scope.
+
+* Any type used in the declaration of a parameter must be one of the allowed
+  types listed below.
+
+* If the function is instantiated from a template, any type used to instantiate
+  the template must be one of the allowed types listed below.
+
+* Uses of function identifiers in the application must assume that the free
+  function kernels are forward declared at the top of the translation unit.
+  Note that this can also affect references to functions that are not declared
+  as free functions kernels as illustrated below.
++
+```
+void foo(int) {/*...*/}
+
+void caller() {
+  auto *pf = foo;  // This is ambiguous because foo(float) is forward declared
+                   // in the integration header
+}
+
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void foo(float) {/*...*/}
+```
+
+The allowed types are:
+
+* A {cpp} fundamental type.
+* A class or struct that is defined at namespace scope.
+* A scoped enumeration that is defined at namespace scope.
+* An unscoped enumeration that has an explicit underlying type, where the
+  enumeration is defined at namespace scope.
+* A type alias to one of the above types.
+
+
+== Example
+
+=== Basic invocation
+
+The following example demonstrates how to define a free function kernel and then
+enqueue it on a device.
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+static constexpr size_t NUM = 1024;
+
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) {
+  // Get the ID of this kernel iteration.
+  size_t id = syclex::this_kernel::get_id();
+
+  ptr[id] = start + static_cast<float>(id);
+}
+
+void main() {
+  sycl::queue q;
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the free function kernel "iota".
+  auto exe_bndl =
+    syclex::get_kernel_bundle<iota, sycl::bundle_state::executable>(ctxt);
+
+  // Get a kernel object for the "iota" function from that bundle.
+  sycl::kernel k_iota = exe_bndl.ext_oneapi_get_kernel<iota>();
+
+  float *ptr = sycl::malloc_shared<float>(NUM, q);
+  q.submit([&](sycl::handler &cgh) {
+    // Set the values of the kernel arguments.
+    cgh.set_args(3.14f, ptr);
+
+    cgh.parallel_for({NUM}, k_iota);
+  }).wait();
+}
+```
+
+=== Free function kernels which are templates or overloaded
+
+A free function kernel may be defined as a function template.
+It is also legal to define several overloads for a free function kernel.
+The following example demonstrates how to get a kernel identifier in such
+cases.
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+template<typename T>
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(T start, T *ptr) {
+  // ...
+}
+
+SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
+void ping(float *x) {
+  // ...
+}
+
+SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
+void ping(int *x) {
+  // ...
+}
+
+int main() {
+  // When the free function kernel is templated, pass the address of a
+  // specific instantiation.
```

Comment:
When the free function kernel is templated, which dtypes are used to instantiate it, since the host code that instantiates it may not be compiled with the kernel code? 

Also, if there are multiple template types (as in the example below), are all allowed combinations of these template types going to be instantiated during the compilation?

`template<typename T, typename V>
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(T start, V *ptr) {
  // ...
}`

Reply:
> When the free function kernel is templated, which dtypes are used to instantiate it, since the host code that instantiates it may not be compiled with the kernel code?

This PR only adds the ability to define a "free function kernel" in a SYCL application, where both host and device code are mixed together in the same source file.  Therefore, the situation you are asking about does not exist.

Are you asking about a case where the kernel is defined as a string and then compiled with the kernel compiler (i.e. "online compiled")?  This is covered in a different PR (#11985).  In particular, see the section titled ["Instantiating templated kernel functions"](https://github.com/intel/llvm/blob/21580c6006012891f1c997bb04fb1c382c5b85fc/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#instantiating-templated-kernel-functions) in that PR.

> Also, if there are multiple template types (as in the example below), are all allowed combinations of these template types going to be instantiated during the compilation?

No, the compiler only instantiates the template with the arguments that you provide.  When the free function kernel is defined in a SYCL application (i.e. not as a string that is online compiled), the application provides the template parameters in the call to `get_kernel_id` as shown in the lines that follow this comment.  Note that the expression `iota<float>` is a normal C++ expression for the address of a a templated function.  By normal C++ rules, this causes the template to be instantiated with the `float` type, so there is no magic SYCL logic here about instantiating free function kernels that are templated.  We're just relying on the normal C++ behavior.

When the kernel is defined as a string, the template is only instantiated with the template parameters that the application provides via the `registered_kernel_names` property.  Again, see the section ["Instantiating templated kernel functions"](https://github.com/intel/llvm/blob/21580c6006012891f1c997bb04fb1c382c5b85fc/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#instantiating-templated-kernel-functions) for details.

Reply:
Thanks for the explanation. I wasn't referring to the case of kernel defined as a string which is online compiled. This means, when the free function kernel is provided in a SYCL source file, the template function must be instantiated from the same source file as shown in the example below.

Reply:
Normal C++ rules require a template to be defined in the same translation unit (i.e. source file) before the template can be instantiated.  If a template needs to be shared across many source files, the normal practice is to define the template in a header file that is included by the different source files.  If the template is instantiated the same way in two or more source files, normal C++ rules will merge them together, effectively choosing just one of the instantiations in the final executable.

None of the behavior about this is specific to SYCL.  It's just the normal rules that apply to C++ templates.